src/sisyphus/cloud.clj

```diff
@@ -269,3 +277,52 @@
             relative-path (.substring remote-key prefix-length)
             local-path (join-path [path relative-path])]
         (download-blob! blob local-path)))))
+
+(defn create-compute-service
+  "Create an instance of com.google.api.services.compute.Compute to make requests with."
+  []
+  (let [transport (GoogleNetHttpTransport/newTrustedTransport)
+        factory (JacksonFactory/getDefaultInstance)
+        auth (into-array ["https://www.googleapis.com/auth/cloud-platform"])
+        credential (GoogleCredential/getApplicationDefault)
+        credential (if (.createScopedRequired credential)
+                     (.createScoped credential (Arrays/asList auth))
+                     credential)
+        builder (Compute$Builder. transport factory credential)]
+    (.setApplicationName builder "Gaia/0.0.1")
+    (.build builder)))
+
+(defn render-filter
+  "Render a map of options into the weird format that the compute instances api expects"
+  [instance-filter]
+  (string/join
+   " "
+   (map
+    (fn [[k v]]
+      (format
+       "(%s = %s)"
+       (name k)
+       (if (string? v)
+         (format "\"%s\"" v)
+         v)))
+    instance-filter)))
+
+(defn list-instances
+  "Given a compute service instance, project and zone for instances, return information on those
+  instances. Also accepts an optional `options` arg that could contain the following keys:
+      * :filter - a map of keys to values to filter the list of instances."
+  ([^Compute service project zone]
+   (list-instances service project zone {}))
+  ([^Compute service project zone options]
+   (let [request (.list (.instances service) project zone)]
+     (if-let [instance-filter (:filter options)]
+       (.setFilter request (render-filter instance-filter)))
+     (loop [response (.execute request)
+            instances []]
+       (let [items (.getItems response)
+             next (.getNextPageToken response)
+             expansion (concat instances items)]
+         (.setPageToken request next)
+         (if (and items next)
+           (recur (.execute request) expansion)
+           expansion))))))
```

Comment:
The API also allows selecting which fields to return. That should speed up the request but I didn't figure out how to express it to the API explorer in a way that worked.

Is there a way in Clojure to declare optional keyword arguments?

Reply:
> The API also allows selecting which fields to return. That should speed up the request but I didn't figure out how to express it to the API explorer in a way that worked.

Perhaps this one? https://developers.google.com/resources/api-libraries/documentation/compute/v1/java/latest/com/google/api/services/compute/Compute.Instances.List.html#setFields-java.lang.String-

Once we know what fields we want we can pare this down. 

> Is there a way in Clojure to declare optional keyword arguments?

There is but it makes applying functions more fussy so I avoid it. Radical simplicity ; )

Reply:
Agreed we should make it work before adding setFields to make it faster.

My previous experiment with setting fields in the API explorer didn't work. The [partial response doc](https://cloud.google.com/compute/docs/api/how-tos/performance#partial-response) gives the key clues (the path syntax for JSON field names). The [instances list API doc](https://cloud.google.com/compute/docs/reference/rest/v1/instances/list) explains the big JSON response structure.

    setFields("items(id,name,hostname,status)")

That works in the API explorer, getting a readable response, and making the API more helpful.

This [link to the API explorer works](https://cloud.google.com/compute/docs/reference/rest/v1/instances/list?apix=true&apix_params=%7B%22project%22%3A%22allen-discovery-center-mcovert%22%2C%22zone%22%3A%22us-west1-b%22%2C%22fields%22%3A%22items(id%2Cname%2Chostname%2Cstatus)%22%7D). It has `fields` set and active but hidden behind `Show standard parameters v`.